### PR TITLE
fix: popup degredaded - WPB-7120

### PR DESCRIPTION
**Pull request by netbe**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-7120" title="WPB-7120" target="_blank">BugWPB-7120</a>  [iOS] "Conversation no longer verified warning" is sent after already sending the message
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Popup for unverified conversation shows up but the message gets sent anyway (first expired then retried automatically)

### Solutions

- Display popup for sending message, before we try sending the message, so we check for the `conversation.mlsVerificationStatus` on each entry point (different type of messages in ConversationViewController).

### Notes

A second PR will deal with the same for popup about calling: incoming calls and outcoming calls

See `E2EIPrivacyWarningChecker` that would do the checks and continue the action after user confirm or cancel the popup.
See `MLSConversationCheckerPresenter` (ConversationViewController+PrivacyAlert.swift:180) to display popup.

Fixed also:
* a legalHold retried message
* CoreData mutlithread issue in Recurring Action

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

### CODE REVIEW

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession+RecurringAction.swift [hunk 1]
```diff
@@ -90,7 +90,8 @@ extension ZMUserSession {
 
             Task {
                 do {
-                    try await self.e2eiRepository.fetchFederationCertificates()
+                    let e2eiRepository = await self.viewContext.perform { self.e2eiRepository }
```
Comment (netbe) [inline wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession+RecurringAction.swift]: fix coredata multithread access

Comment (theswiftway) [reply wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession+RecurringAction.swift]: **question:** Is there a way we could have a precondition to check if these core data values are accessed on the correct context?

Comment (caldrian) [reply wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession+RecurringAction.swift]: suggestion: try to get everything done with only one `context.perform`:

```
var refreshFederationCertificatesAction: RecurringAction {
    .init(id: #function, interval: .oneDay) { [weak self] in
        Task { [weak self] in
            do {
                guard let self else { return }
                let (e2eiFeature, e2eiRepository) = await viewContext.perform { (self.e2eiFeature, self.e2eiRepository) }
                guard e2eiFeature.isEnabled else { return }

                try await e2eiRepository.fetchFederationCertificates()
            } catch {
                WireLogger.e2ei.error("Failed to refresh federation certificates: \(error)")
            }
        }
    }
}
```

File: wire-ios/Wire-iOS/Sources/UserInterface/Calling/CallController/CallController.swift [hunk 2]
```diff
@@ -130,8 +130,27 @@ extension CallController: WireCallCenterCallStateObserver {
                              timestamp: Date?,
                              previousCallState: CallState?) {
         presentUnsupportedVersionAlertIfNecessary(callState: callState)
-        presentSecurityDegradedAlertIfNecessary(for: conversation.voiceChannel)
-        updateActiveCallPresentationState()
+
+        // TODO: how to detect this is the first time we get the call sa degraded
+        if case .incoming(_, shouldRing: true, degraded: true) = callState {
```
Comment (netbe) [inline wire-ios/Wire-iOS/Sources/UserInterface/Calling/CallController/CallController.swift]: this checks work, but I wonder if there's a better place to check as the method will be called multiple times resulting in showing this popup multiple times

File: wire-ios-data-model/Source/Model/Message/ZMMessage+Insert.swift [hunk 3]
```diff
@@ -48,6 +49,18 @@ extension ZMMessage {
...
+        visibleInConversation = hiddenInConversation
+        hiddenInConversation = nil
+    }
+
+    func hideMessage() {
+        guard visibleInConversation != nil else { return }
+        hiddenInConversation = visibleInConversation
+        visibleInConversation = nil
+    }
+
```
Comment (netbe) [inline wire-ios-data-model/Source/Model/Message/ZMMessage+Insert.swift]: not used

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession+RecurringAction.swift (hunk 1 above)
Comment (netbe) [reply wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession+RecurringAction.swift]: replaced

**Review by typfel — APPROVED** [review]
Ah nice so we are check before we even ask the user session to send a message, makes sense.

**Review by typfel — APPROVED** [review]

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/ConversationViewController+PrivacyAlert.swift [hunk 4]
```diff
@@ -70,6 +70,11 @@ extension ConversationViewController {
             return
         }
 
+        presentAlert(with: alertContent)
+    }
+
+    private func presentAlert(with alertContent: AlertContent) {
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/ConversationViewController+PrivacyAlert.swift:76]: 👍

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/InputBar/E2EIPrivacyWarningChecker.swift [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    static func e2eiPrivacyWarningConfirm(sendAnyway: Bool) {
+        NotificationCenter.default.post(.e2eiPrivacyWarningConfirm(sendAnyway: sendAnyway))
+    }
+
+    // add object in charge to present e2eiPrivacyWarningAlert
+    static func addPresenter(_ observer: E2EIPrivacyWarningPresenter) -> NSObjectProtocol {
+       return NotificationCenter.default.addObserver(forName: .presentE2EIPrivacyWarningAlert, object: nil, queue: .main) { note in
+            observer.presentE2EIPrivacyWarningAlert(note)
+        }
+    }
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/InputBar/E2EIPrivacyWarningChecker.swift]: nitpick:
- `return` can be omitted
- indentation is off

File: wire-ios/Wire-iOS/Sources/UserInterface/Conversation/InputBar/E2EIPrivacyWarningChecker.swift [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+extension Notification {
+
+    var alertType: E2EIPrivacyWarningChecker.AlertType? {
+        userInfo?["alertType"] as? E2EIPrivacyWarningChecker.AlertType
+    }
+}
+private extension Notification {
```
Comment (markusfassbender) [inline wire-ios/Wire-iOS/Sources/UserInterface/Conversation/InputBar/E2EIPrivacyWarningChecker.swift]:
```suggestion
}

private extension Notification {
```

**Review by markusfassbender — APPROVED** [review]

**Review by KaterinaWire — APPROVED** [review]

**Pull request merged**